src/wifipaf/WiFiPAFTP.cpp

```diff
@@ -0,0 +1,607 @@
+                     err = WIFIPAF_ERROR_INVALID_PAFTP_HEADER_FLAGS);
+
+        // Add received fragment to reassembled message buffer.
+        mRxBuf->AddToEnd(std::move(data));
+        mRxBuf->CompactHead(); // will free 'data' and adjust rx buf's end/length
+
+        // For now, limit WiFiPAFTP message size to max length of 1 pbuf, as we do for chip messages sent via IP.
+        // TODO add support for WiFiPAFTP messages longer than 1 pbuf
+        VerifyOrExit(!mRxBuf->HasChainedBuffer(), err = CHIP_ERROR_INBOUND_MESSAGE_TOO_BIG);
+    }
+    else
+    {
+        err = WIFIPAF_ERROR_REASSEMBLER_INCORRECT_STATE;
+        ExitNow();
+    }
+
+    if (rx_flags.Has(HeaderFlags::kEndMessage))
+    {
+        // Trim remainder, if any, of the received packet buffer based on sender-specified length of reassembled message.
+        VerifyOrExit(CanCastTo<uint16_t>(mRxBuf->DataLength()), err = CHIP_ERROR_MESSAGE_TOO_LONG);
+        int padding = static_cast<uint16_t>(mRxBuf->DataLength()) - mRxLength;
+
+        if (padding > 0)
+        {
+            mRxBuf->SetDataLength(static_cast<size_t>(mRxLength));
+        }
+
+        // Ensure all received fragments add up to sender-specified total message size.
+        VerifyOrExit(mRxBuf->DataLength() == mRxLength, err = WIFIPAF_ERROR_REASSEMBLER_MISSING_DATA);
+
+        // We've reassembled the entire message.
+        mRxState = kState_Complete;
+        mRxPacketCount++;
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        mRxState = kState_Error;
+        // Dump protocol engine state, plus header flags and received data length.
+        ChipLogError(WiFiPAF, "HandleCharacteristicReceived failed, err = %" CHIP_ERROR_FORMAT ", rx_flags = %u", err.Format(),
+                     rx_flags.Raw());
+        if (didReceiveAck)
+        {
+            ChipLogError(WiFiPAF, "With rx'd ack = %u", receivedAck);
+        }
+        if (!mRxBuf.IsNull())
+        {
+            ChipLogError(WiFiPAF, "With rx buf data length = %u", static_cast<unsigned>(mRxBuf->DataLength()));
+        }
+        LogState();
+
+        if (!data.IsNull()) // NOLINT(bugprone-use-after-move)
+        {
+            // Tack received data onto rx buffer, to be freed when end point resets protocol engine on close.
+            if (!mRxBuf.IsNull())
+            {
+                mRxBuf->AddToEnd(std::move(data));
+            }
+            else
+            {
+                mRxBuf = std::move(data);
+            }
+        }
+    }
+
+    return err;
+}
+
+PacketBufferHandle WiFiPAFTP::TakeRxPacket()
+{
+    if (mRxState == kState_Complete)
+    {
+        mRxState = kState_Idle;
+    }
+    return std::move(mRxBuf);
+}
+
+// Calling convention:
+//   May only be called if data arg is committed for immediate, synchronous subsequent transmission.
+//   Returns false on error. Caller must free data arg on error.
+bool WiFiPAFTP::HandleCharacteristicSend(System::PacketBufferHandle data, bool send_ack)
```

Comment:
Again, what's with "characteristic"?

Reply:
It should be the follow-up packets. I will change it